### ISSUE TITLE
Proposal to Update REST API

### ISSUE THREAD

**Issue by someoneAnyone** [nightscout/cgm-remote-monitor#344]
Per the conversation in gitter.im:

There are multiple clients (watches too) working their way through development that leverage nightscout data. It would be great if more of the data was available via rest APIs so things can be consistently and generically handled.

Can the APIs be modified so that items like bg-delta, or prediction entries are available?
##

**Comment by someoneAnyone** [comment]
Just adding some interesting tools/documentation i've encountered that might be useful for this feature/project. Has the teem considered using something like: http://loopback.io for documenting and exposing the APIs? might help in testing too.

**Comment by jimsiff** [comment]
It would also be nice to expose any mbgs, filtered, unfiltered, noise, and rssi data at some point.

**Comment by jasoncalabrese** [comment]
all that is getting filtered by the sgvdata module, I think we're going to pull it out.  Right now the api only supports the count param, really need some pagination, maybe query by date

**Comment by someoneAnyone** [comment]
Here are some nice tips for REST design. http://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api#requirements

**Comment by bewest** [comment]
@jimsiff expressed some interest in helping document the API.  If possible, there's a tool called swagger that might help a lot.

I modified the example uber api swagger spec to come up with this one: https://gist.github.com/bewest/5fb2910aa09a94052712

You can drag it/import it into swagger's demo API and it generates very nice documentation, complete with the objects to expect in the responses: http://editor.swagger.io/#/

![image](https://cloud.githubusercontent.com/assets/394179/6716488/1f04bf74-cd63-11e4-8e6b-6cbf85204017.png)

**Comment by bewest** [comment]
Swagger has some nifty tools, one of them can help visualize the api and resources as an interactive graphic:
![image](https://cloud.githubusercontent.com/assets/394179/6718119/9b3015f2-cd6e-11e4-92e0-b5a1f36b6645.png)

**Comment by someoneAnyone** [comment]
It can also generate client code which is nice.

**Comment by bewest** [comment]
@someoneAnyone hehe, patches welcome, hehehe.

Turns out the REST API is already capable of some fairly advanced searching and filtering, if you know the right syntax: https://github.com/nightscout/cgm-remote-monitor/pull/504  I'll be working on extending it with feedback from community.

I updated that gist with a slightly more complete data model, HTH.

**Comment by bewest** [comment]
@jimsiff I created this repo https://github.com/bewest/ns-types where I've combined a lot of notes taken over the last year.

I'd like to use this to create website devoted to just discussing the data model, and have it available as a module for NS.  My hope is this will replace sgvdata by including definitions for all the data types.  Since it's got all the data-types, it might be a good place to document the "design" of the API as well, or perhaps it better in the wiki in this repo?  Thoughts/feedback?
